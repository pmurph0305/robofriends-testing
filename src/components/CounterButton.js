import React, { Component } from 'react';

class CounterButton extends Component {

    constructor() {
        super();
        this.state = {
            count: 0,
        }
    }

    // Called before render(). Use cautiously, measure performance before using.
    shouldComponentUpdate(nextProps, nextState) {
        // console.log("NextProps", nextProps);
        // console.log("NextState", nextState);
        // Even though parent is being re-rendered, this would prevent this component from rerendering.
        // Does essentially the same thing as extending PureComponent.
        // Except it uses shallow comparison, so it can miss deep nested component updates.
        if (this.state.count !== nextState.count) {
            return true;
        } else {
            return false;
        }
    }

    updateCount = () => {
        this.setState(state => { 
            return { count: this.state.count + 1 }
        })
    }

    render() {
        return(
            <button color={this.props.color} onClick={this.updateCount} id='counter'>
                Count: {this.state.count}
            </button>
        );
    }
}

export default CounterButton;