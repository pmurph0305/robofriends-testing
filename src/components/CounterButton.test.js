import { shallow } from 'enzyme';
import React from 'react';
import CounterButton from './CounterButton';

describe('My own coverage tests from Exercise #2', () => {
// The following is from the exercise to get the count button up to 100% code coverage.
    it('expect to render CounterButton Component', () => {
        expect(shallow(<CounterButton />)).toMatchSnapshot();
    })

    // testing state by simulating a click.
    it('simulate click events on Counter Button', () => {
        const wrapper = shallow(<CounterButton/>);
        wrapper.find('button').simulate('click');
        expect(wrapper.state('count')).toEqual(1);
    })

    // After seeing course was using older non-extend PureComponent but regular component with shouldComponentUpdate
    // I made these changes and branch coverage % dropped to 50% coverage.
    // So I paused video, and created the changes to regular Component and covered the branch differences.
    it('test should component update on Counter Button', () => {
        const wrapper = shallow(<CounterButton/>);
        const shouldCompUpdateFalse = wrapper.instance().shouldComponentUpdate(null, { count: 0 })
        expect(shouldCompUpdateFalse).toBe(false);
        const shouldCompUpdateTrue = wrapper.instance().shouldComponentUpdate(null, { count: 1 })
        expect(shouldCompUpdateTrue).toBe(true);
    })
})

describe('Examples from the course videos', () => {
    // Tests from the video.
    it('Course expect to render CounterButton Component', () => {
        const mockColor = 'red'
        expect(shallow(<CounterButton color={mockColor}/>)).toMatchSnapshot();
    })

    it('Course correctly increments counter', () => {
        const mockColor = 'red';
        const wrapper = shallow(<CounterButton color={mockColor}/>);
        wrapper.find('[id="counter"]').simulate('click');
        wrapper.find('[id="counter"]').simulate('click');
        expect(wrapper.state()).toEqual({ count: 2})
        wrapper.find('[id="counter"]').simulate('click');
        expect(wrapper.state()).toEqual({ count: 3})
        wrapper.find('[id="counter"]').simulate('keypress');
        expect(wrapper.state()).toEqual({ count: 3})
        expect(wrapper.props().color).toEqual('red')
    })
})