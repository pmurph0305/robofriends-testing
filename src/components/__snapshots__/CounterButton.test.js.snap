// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Examples from the course videos Course expect to render CounterButton Component 1`] = `ShallowWrapper {}`;

exports[`My own coverage tests from Exercise #2 expect to render CounterButton Component 1`] = `ShallowWrapper {}`;
